dm/pkg/checker/table_structure.go

```diff
@@ -296,96 +331,137 @@ func (c *ShardingTablesChecker) Check(ctx context.Context) *Result {
-					r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.name)
-					return r
-				}
+	concurrency, err := getConcurrency(ctx, c.tableMap, c.dbs, c.dumpThreads)
+	if err != nil {
+		markCheckError(r, err)
+		return r
+	}
+	eg, checkCtx := errgroup.WithContext(ctx)
+	for i := 0; i < concurrency; i++ {
+		eg.Go(func() error {
+			return c.checkShardingTable(checkCtx, r)
+		})
+	}
 
-				ctStmt, ok := stmt.(*ast.CreateTableStmt)
-				if !ok {
-					markCheckError(r, errors.Errorf("Expect CreateTableStmt but got %T", stmt))
-					r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.name)
-					return r
-				}
+	dispatchTableItem(checkCtx, c.tableMap, c.inCh)
+	close(c.inCh)
+	if err := eg.Wait(); err != nil {
+		c.reMu.Lock()
+		markCheckError(r, err)
+		c.reMu.Unlock()
+	}
 
-				if c.checkAutoIncrementPrimaryKey {
-					passed := c.checkAutoIncrementKey(instance, schema, table, ctStmt, info, r)
-					if !passed {
-						return r
-					}
+	log.L().Logger.Info("check sharding table structure over", zap.String("spend time", time.Since(startTime).String()))
+	return r
+}
+
+func (c *ShardingTablesChecker) checkShardingTable(ctx context.Context, r *Result) error {
+	var (
+		sourceID string
+		p        *parser.Parser
+		err      error
+	)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case checkItem, ok := <-c.inCh:
+			if !ok {
+				return nil
+			}
+			table := checkItem.table
+			if len(sourceID) == 0 || sourceID != checkItem.sourceID {
+				sourceID = checkItem.sourceID
+				p, err = dbutil.GetParserForDB(ctx, c.dbs[sourceID])
+				if err != nil {
+					c.reMu.Lock()
+					r.Extra = fmt.Sprintf("fail to get parser for sourceID %s on sharding %s", sourceID, c.targetTableID)
+					c.reMu.Unlock()
+					return err
 				}
+			}
 
-				if stmtNode == nil {
-					stmtNode = ctStmt
-					firstTable = dbutil.TableName(schema, table)
-					firstInstance = instance
+			statement, err := dbutil.GetCreateTableSQL(ctx, c.dbs[sourceID], table.Schema, table.Name)
+			if err != nil {
+				// continue if table was deleted when checking
+				if isMySQLError(err, mysql.ErrNoSuchTable) {
 					continue
 				}
+				return err
+			}
 
-				checkErr := c.checkConsistency(stmtNode, ctStmt, firstTable, dbutil.TableName(schema, table), firstInstance, instance)
-				if checkErr != nil {
-					r.State = StateFailure
-					r.Errors = append(r.Errors, checkErr)
-					r.Extra = fmt.Sprintf("error on sharding %s", c.name)
-					r.Instruction = "please set same table structure for sharding tables"
-					return r
+			info, err := dbutil.GetTableInfoBySQL(statement, p)
+			if err != nil {
+				return errors.Annotatef(err, "statement: %s", statement)
+			}
+
+			ctStmt, err := getCreateTableStmt(p, statement)
+			if err != nil {
+				return err
+			}
+
+			if c.checkAutoIncrementPrimaryKey {
+				passed := c.checkAutoIncrementKey(sourceID, table.Schema, table.Name, ctStmt, info, r)
+				if !passed {
+					return nil
```

Comment:
the beheavior of `checkAutoIncrementKey` and `checkConsistency` is  not same, checkAutoIncrementKey handle check result inner but checkConsistency not, this confused me when i see 
```
if !passed {
  return nil
}
```
could please adjust this two checker? maybe move 
```
		if !hasMatchedRule {
			r.State = StateFailure
			r.Errors = append(r.Errors, NewError("sourceID %s table `%s`.`%s` of sharding %s have auto-increment key %s and column mapping, but type of %s should be bigint", sourceID, schema, table, c.targetTableID, columnName, columnName))
			r.Instruction = "please handle it by yourself"
			r.Extra = AutoIncrementKeyChecking
			return false
		}
```
to outer like checkConsistency